pipelines/src/main/java/au/org/ala/pipelines/beam/ALAUUIDMintingPipeline.java

```diff
@@ -0,0 +1,247 @@
+     *
+     * @param source
+     * @param uniqueDwcTerms
+     * @return
+     * @throws RuntimeException
+     */
+    public static String generateUniqueKey(ExtendedRecord source, List<DwcTerm> uniqueDwcTerms) throws RuntimeException {
+        List<String> uniqueValues = new ArrayList<String>();
+        for (DwcTerm dwcTerm : uniqueDwcTerms) {
+            String value = ModelUtils.extractNullAwareValue(source, dwcTerm);
+            if (value == null || StringUtils.trimToNull(value) == null) {
+                throw new RuntimeException("Unable to load dataset. Unique term empty for record term: " + dwcTerm);
+            }
+            uniqueValues.add(value.trim());
+        }
+        //create the unique key
+        return String.join("|", uniqueValues);
+    }
+
+    /**
+     * Function to create ALAUUIDRecords.
+     */
+    static class CreateALAUUIDRecordFcn extends DoFn<KV<String, KV<String, String>>, ALAUUIDRecord> {
+
+        private Counter orphanedUniqueKeys = Metrics.counter(CreateALAUUIDRecordFcn.class, "orphanedUniqueKeys");
+        private Counter newUuids = Metrics.counter(CreateALAUUIDRecordFcn.class, "newUuids");
+        private Counter preservedUuids = Metrics.counter(CreateALAUUIDRecordFcn.class, "preservedUuids");
+
+        @ProcessElement
+        public void processElement(@Element KV<String, KV<String, String>> uniqueKeyMap, OutputReceiver<ALAUUIDRecord> out) {
+
+            // get the constructed key
+            String uniqueKey = uniqueKeyMap.getKey();
+
+            //get the matched ExtendedRecord.getId()
+            String id = uniqueKeyMap.getValue().getKey();
+
+            //get the UUID
+            String uuid = uniqueKeyMap.getValue().getValue();
+
+            //if UUID == NO_ID_MARKER, we have a new record so we need a new UUID.
+            if (uuid.equals(NO_ID_MARKER)){
+                newUuids.inc();
+                uuid = UUID.randomUUID().toString();
+            } else if(id.equals(NO_ID_MARKER)){
+                orphanedUniqueKeys.inc();
+            } else {
+                preservedUuids.inc();
+            }
+
+            ALAUUIDRecord aur = ALAUUIDRecord.newBuilder().setUniqueKey(uniqueKey).setUuid(uuid).setId(id).build();
+            out.output(aur);
+        }
+    }
+
+    /**
+     * Transform to create a map of unique keys built from previous runs and UUID.
+     */
+    static class ALAUUIDRecordKVFcn extends DoFn<ALAUUIDRecord, KV<String, String>> {
+        @ProcessElement
+        public void processElement(@Element ALAUUIDRecord alaUUIDRecord, OutputReceiver<KV<String, String>> out) {
+            out.output(KV.of(alaUUIDRecord.getUniqueKey(), alaUUIDRecord.getUuid()));
+        }
+    }
+
+    /**
+     * Match the darwin core term which has been supplied in simple camel case format e.g. catalogNumber.
+     *
+     * @param name
+     * @return
+     */
+    static Optional<DwcTerm> getDwcTerm(String name){
```

Comment:
This method needs to be removed, and the terms need to be matched directly, not only if they are recognised Darwin Core Terms.

CC @peggynewman @RobinaSanderson @sadeghim